src/Microsoft.Health.Fhir.Api/Features/Audit/AuditHeaderReader.cs

```diff
@@ -0,0 +1,57 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+
+namespace Microsoft.Health.Fhir.Api.Features.Audit
+{
+    public class AuditHeaderReader : IAuditHeaderReader
+    {
+        private readonly AuditConfiguration _auditConfiguration;
+
+        public AuditHeaderReader(IOptions<AuditConfiguration> auditConfiguration)
+        {
+            _auditConfiguration = auditConfiguration.Value;
+        }
+
+        public IReadOnlyDictionary<string, string> Read(HttpContext httpContext)
+        {
+            if (!httpContext.Items.ContainsKey(AuditConstants.CustomAuditHeaderKeyValue))
+            {
+                var customHeaders = new Dictionary<string, string>();
+
+                foreach (var headerName in httpContext.Request.Headers.Keys)
```

Comment:
nit: if you do `foreach(KeyValuePair<string, StringValues> header in httpContext.Request.Headers)`, then you can access the key and the value without additional lookup.